src/category_theory/subobject.lean

```diff
@@ -0,0 +1,1330 @@
+thin_skeleton.from_thin_skeleton _
+
+
+/--
+Starting with `A : mono_over X`, we can take its equivalence class in `subobject X`
+then pick an arbitrary representative using `representative.obj`.
+This is isomorphic (in `mono_over X`) to the original `A`.
+-/
+noncomputable
+def representative_iso {X : C} (A : mono_over X) :
+  representative.obj ((to_thin_skeleton _).obj A) ≅ A :=
+(thin_skeleton.from_thin_skeleton _).as_equivalence.counit_iso.app A
+
+/--
+Use choice to pick a representative underlying object in `C` for any `subobject X`.
+
+Prefer to use the coercion `P : C` rather than explicitly writing `underlying.obj P`.
+-/
+noncomputable
+def underlying {X : C} : subobject X ⥤ C :=
+representative ⋙ mono_over.forget _ ⋙ over.forget _
+
+instance : has_coe (subobject X) C :=
+{ coe := λ Y, underlying.obj Y, }
+
+@[simp] lemma underlying_as_coe {X : C} (P : subobject X) : underlying.obj P = P := rfl
+
+/--
+If we construct a `subobject Y` from an explicit `f : X ⟶ Y` with `[mono f]`,
+then pick an arbitrary choice of underlying object `(subobject.mk f : C)` back in `C`,
+it is isomorphic (in `C`) to the original `X`.
+-/
+noncomputable
+def underlying_iso {X Y : C} (f : X ⟶ Y) [mono f] : (subobject.mk f : C) ≅ X :=
+(mono_over.forget _ ⋙ over.forget _).map_iso (representative_iso (mono_over.mk' f))
+
+/--
+The morphism in `C` from the arbitrarily chosen underlying object to the ambient object.
+-/
+noncomputable
+def arrow {X : C} (Y : subobject X) : (Y : C) ⟶ X :=
+(representative.obj Y).val.hom
+
+instance arrow_mono {X : C} (Y : subobject X) : mono (Y.arrow) :=
+(representative.obj Y).property
+
+@[simp]
+lemma underlying_arrow {X : C} {Y Z : subobject X} (f : Y ⟶ Z) :
+  underlying.map f ≫ arrow Z = arrow Y :=
+over.w (representative.map f)
+
+@[simp]
+lemma underlying_iso_arrow {X Y : C} (f : X ⟶ Y) [mono f] :
+  (underlying_iso f).inv ≫ (subobject.mk f).arrow = f :=
+over.w _
+
+/-- Two morphisms into a subobject are equal exactly if
+the morphisms into the ambient object are equal -/
+@[ext]
+lemma eq_of_comp_arrow_eq {X Y : C} {P : subobject Y}
+  {f g : X ⟶ P} (h : f ≫ P.arrow = g ≫ P.arrow) : f = g :=
+(cancel_mono P.arrow).mp h
+
+-- TODO surely there is a cleaner proof here
+lemma le_of_comm {B : C} {X Y : subobject B} (f : (X : C) ⟶ (Y : C)) (w : f ≫ Y.arrow = X.arrow) :
+  X ≤ Y :=
+begin
+  revert f w,
+  refine quotient.induction_on₂' X Y _,
+  intros P Q f w,
+  fsplit,
+  refine over.hom_mk ((representative_iso P).inv.left ≫ f ≫ (representative_iso Q).hom.left) _,
+  dsimp,
+  simp only [over.w, category.assoc],
+  erw [w, (representative_iso P).inv.w],
+  dsimp,
+  simp only [category.comp_id],
+end
+
+/-- When `f : X ⟶ Y` and `P : subobject Y`,
+`P.factors f` expresses that there exists a factorisation of `f` through `P`.
+Given `h : P.factors f`, you can recover the morphism as `P.factor_thru f h`.
+-/
+def factors {X Y : C} (P : subobject Y) (f : X ⟶ Y) : Prop := ∃ g : X ⟶ P, g ≫ P.arrow = f
```

Comment:
I feel like this is better off defined in terms of `quotient.lift_on` instead of picking a representative - there are cases when I wanted to use this predicate for a subobject constructed by `subobject.mk` so it would be nice to have that definitionally reduce to something nicer rather than needing to insert isomorphisms ad-hoc

Reply:
Is https://github.com/leanprover-community/mathlib/pull/6302/commits/9f11a5a24850017f43c565c6bcb8a6e32c361b58 what you had in mind?

It came at a cost of inserting `(factors_iff _ _).mpr` at a bunch of places later.

Reply:
Yeah, exactly.

> It came at a cost of inserting (factors_iff _ _).mpr at a bunch of places later.

In that case can you have a named lemma which gives this, and use that lemma instead? I suspect some of these proofs can be golfed anyway...

Reply:
As a concrete example, you can change the proof of `top_factors` (for subobjects) to:
```
lemma top_factors {A B : C} (f : A ⟶ B) : (⊤ : subobject B).factors f :=
⟨f, comp_id _⟩
```
so the isomorphism doesn't need to be there at all (and no mpr). That said, probably there should be `mono_over.top_factors`, and the story of proving things for `mono_over` and deducing them in one line for `subobject` happens again

Reply:
I went through and made more of this sort of change, it's in the branch `subobject-lifts-quotient`. In my opinion the proofs are nicer but that's subjective. To me this is a bit like the `has_limit` vs `is_limit` story in limits - we can prove things about the limit which `choice` gives for `has_limit`, and then translate along isos to get it for other limits, but it feels nicer to prove it about any limit and deduce it for "the" limit easily

Reply:
Great, thank you! I've merged your proposals.